ide/ide_slave.ml

```diff
@@ -78,56 +79,42 @@ let is_known_option cmd = match cmd with
   | VernacUnsetOption o -> coqide_known_option o
   | _ -> false
 
-let is_debug cmd = match cmd with
-  | VernacSetOption (["Ltac";"Debug"], _) -> true
-  | _ -> false
-
-let is_query cmd = match cmd with
-  | VernacChdir None
-  | VernacMemOption _
-  | VernacPrintOption _
-  | VernacCheckMayEval _
-  | VernacGlobalCheck _
-  | VernacPrint _
-  | VernacSearch _
-  | VernacLocate _ -> true
-  | _ -> false
-
-let is_undo cmd = match cmd with
-  | VernacUndo _ | VernacUndoTo _ -> true
-  | _ -> false
-
 (** Check whether a command is forbidden by CoqIDE *)
 
 let coqide_cmd_checks (loc,ast) =
   let user_error s = CErrors.user_err ~loc ~hdr:"CoqIde" (str s) in
   if is_debug ast then
     user_error "Debug mode not available within CoqIDE";
   if is_known_option ast then
-    Feedback.msg_warning (strbrk"This will not work. Use CoqIDE view menu instead");
-  if Vernac.is_navigation_vernac ast || is_undo ast then
-    Feedback.msg_warning (strbrk "Rather use CoqIDE navigation instead");
+    Feedback.msg_warning ~loc (strbrk"This will not work. Use CoqIDE view menu instead");
+  if is_navigation_vernac ast || is_undo ast then
+    Feedback.msg_warning ~loc (strbrk "Rather use CoqIDE navigation instead");
   if is_query ast then
-    Feedback.msg_warning (strbrk "Query commands should not be inserted in scripts")
+    Feedback.msg_warning ~loc (strbrk "Query commands should not be inserted in scripts")
 
 (** Interpretation (cf. [Ide_intf.interp]) *)
 
 let add ((s,eid),(sid,verbose)) =
-  let newid, rc = Stm.add ~ontop:sid verbose ~check:coqide_cmd_checks eid s in
-  let rc = match rc with `NewTip -> CSig.Inl () | `Unfocus id -> CSig.Inr id in
+  let pa      = Pcoq.Gram.parsable (Stream.of_string s) in
+  let loc_ast = Stm.parse_sentence sid pa               in
+  coqide_cmd_checks loc_ast;
+  let newid, rc = Stm.add ~ontop:sid verbose loc_ast in
+  let rc        = match rc with `NewTip -> CSig.Inl () | `Unfocus id -> CSig.Inr id in
```

Comment:
I'll digress a bit, but I've been influenced lately by the radical view of Elm w.r.t indenting style (cf. http://elm-lang.org/docs/style-guide). They even have an automatic formatter on save (https://github.com/avh4/elm-format) to avoid thinking about it and keep a consistent style within a project and even across projects. I have the feeling that there are some good ideas to borrow.